app/models/tube.rb

```diff
@@ -84,19 +84,32 @@ def after_comment_addition(comment)
   end
 
   def self.create_with_barcode!(*args, &block)
-    attributes = args.extract_options!
-    barcode = args.first || attributes.delete(:barcode)
-    prefix = attributes.delete(:barcode_prefix)&.prefix || default_prefix
-    if barcode.present?
-      human = SBCF::SangerBarcode.new(prefix: prefix, number: barcode).human_barcode
-      raise "Barcode: #{barcode} already used!" if Barcode.exists?(barcode: human)
-    end
+    attributes = args.extract_options!.symbolize_keys
+
+    barcode, prefix = extract_barcode(args, attributes)
+    validate_barcode(barcode, prefix) if barcode.present?
     barcode ||= AssetBarcode.new_barcode
-    primary_barcode = { prefix: prefix, number: barcode }
-    create!(attributes.merge(sanger_barcode: primary_barcode), &block)
+
+    foreign_barcode = attributes.delete(:foreign_barcode)
```

Comment:
Why are you deleting the attribute, only to immediately set it?
Create will quite happily use the setter you've created if it sees the right attribute.
If its a case of << not working until after the tube it persisted, I believe you can use barcodes.build(...) instead and it will defer attempting to save the barcode until the parent record is persisted.

Reply:
Hi James, I had some trouble getting the barcodes to be added in the correct order - so, the sanger one first, and then the foreign one, to ensure that the foreign one becomes the primary. This way I can control the order explicitly.

Reply:
Ahh, that makes some sense. I'm wondering if we *need* to generate a Sanger barcode for everything. The fact we do was more an accident of design, followed by a "sure, why not".

Reply:
Ah fair enough, I didn't know the history! But sure, I can't think of a good reason why we need a Sanger barcode.